File: benchmarks/bench_regex.pl
#!/usr/bin/env perl
# Perl counterpart of bench_regex.strada (identical workload).
use strict; use warnings; use Time::HiRes qw(time);

my @lines;
for my $i (0..199999) {
    my $code = 200 + (($i * 7) % 4) * 100;
    my $bytes = ($i * 37) % 100000;
    push @lines, sprintf('10.0.%d.%d - - [12/Jun/2026:01:0%d:00] "GET /page/%d HTTP/1.1" %d %d',
        int($i/256) % 256, $i % 256, $i % 10, $i % 1000, $code, $bytes);
}

my $t0 = time;
my $hits = 0;
for (@lines) { $hits++ if /" (4|5)\d\d / }
my $t1 = time;
printf "match: %d %.6f\n", $hits, $t1 - $t0;

my $bytes_total = 0;
for (@lines) { $bytes_total += $4 if /^(\S+) .* "(\w+) [^"]*" (\d+) (\d+)$/ }
my $t2 = time;
printf "captures: %d %.6f\n", $bytes_total, $t2 - $t1;

my $code_sum = 0;
for my $i (0..99999) { $code_sum += $+{code} if $lines[$i] =~ /" (?<code>\d+) (?<bytes>\d+)$/ }
my $t3 = time;
printf "named: %d %.6f\n", $code_sum, $t3 - $t2;

my $sub_len = 0;
for my $i (0..99999) { my $l = $lines[$i]; $l =~ s/\d+/N/g; $sub_len += length $l }
my $t4 = time;
printf "subst: %d %.6f\n", $sub_len, $t4 - $t3;

my $sube_len = 0;
for my $i (0..49999) { my $l = $lines[$i]; $l =~ s/(\d+)/length($1)/eg; $sube_len += length $l }
my $t5 = time;
printf "subst-e: %d %.6f\n", $sube_len, $t5 - $t4;

my $tr_count = 0;
for my $i (0..99999) { my $l = $lines[$i]; $tr_count += ($l =~ tr/a-z/A-Z/) }
my $t6 = time;
printf "tr: %d %.6f\n", $tr_count, $t6 - $t5;
printf "total: %.6f\n", $t6 - $t0;

File: benchmarks/bench_regex.js
#!/usr/bin/env node
// Node counterpart of bench_regex.strada (identical workload; tr/// is
// emulated with replace+callback, the closest JS equivalent).
const lines = [];
for (let i = 0; i < 200000; i++) {
    const code = 200 + ((i * 7) % 4) * 100;
    const bytes = (i * 37) % 100000;
    lines.push(`10.0.${Math.floor(i/256)%256}.${i%256} - - [12/Jun/2026:01:0${i%10}:00] "GET /page/${i%1000} HTTP/1.1" ${code} ${bytes}`);
}

let t0 = Date.now();
let hits = 0;
const reMatch = /" (4|5)\d\d /;
for (const l of lines) if (reMatch.test(l)) hits++;
let t1 = Date.now();
console.log("match:", hits, (t1 - t0) / 1000);

let bytesTotal = 0;
const reCap = /^(\S+) .* "(\w+) [^"]*" (\d+) (\d+)$/;
for (const l of lines) { const m = reCap.exec(l); if (m) bytesTotal += +m[4]; }
let t2 = Date.now();
console.log("captures:", bytesTotal, (t2 - t1) / 1000);

let codeSum = 0;
const reNamed = /" (?<code>\d+) (?<bytes>\d+)$/;
for (let i = 0; i < 100000; i++) { const m = reNamed.exec(lines[i]); if (m) codeSum += +m.groups.code; }
let t3 = Date.now();
console.log("named:", codeSum, (t3 - t2) / 1000);

let subLen = 0;
for (let i = 0; i < 100000; i++) subLen += lines[i].replace(/\d+/g, "N").length;
let t4 = Date.now();
console.log("subst:", subLen, (t4 - t3) / 1000);

let subeLen = 0;
for (let i = 0; i < 50000; i++) subeLen += lines[i].replace(/(\d+)/g, (m) => String(m.length)).length;
let t5 = Date.now();
console.log("subst-e:", subeLen, (t5 - t4) / 1000);

let trCount = 0;
for (let i = 0; i < 100000; i++) {
    let c = 0;
    lines[i].replace(/[a-z]/g, (m) => { c++; return m.toUpperCase(); });
    trCount += c;
}
let t6 = Date.now();
console.log("tr:", trCount, (t6 - t5) / 1000);
console.log("total:", (t6 - t0) / 1000);
